File: server/src/services/ai.service.ts
import axios from 'axios';
import logger from '../utils/logger';

interface TrafficSituation {
  locationId: string;
  currentCongestion: string;
  vehicleCount: number;
  averageSpeed: number;
  timestamp: Date;
}

interface AIAnalysisResult {
  analysis: string;
  recommendations: string[];
  severity: string;
}

export class AIService {
  private apiKey: string;
  private baseURL: string;

  constructor() {
    this.apiKey = process.env.FEATHERLESS_API_KEY || '';
    this.baseURL = 'https://api.featherless.ai/v1';
  }

  async analyzeTrafficSituation(situation: TrafficSituation): Promise<AIAnalysisResult> {
    try {
      const prompt = `Analyze this traffic situation and provide recommendations:
Location: ${situation.locationId}
Current Congestion: ${situation.currentCongestion}
Vehicle Count: ${situation.vehicleCount}
Average Speed: ${situation.averageSpeed} km/h
Time: ${situation.timestamp.toISOString()}

Provide:
1. Brief analysis of the situation
2. 3-5 actionable recommendations for drivers
3. Severity assessment (low/moderate/high/critical)`;

      const response = await axios.post(
        `${this.baseURL}/chat/completions`,
        {
          model: 'meta-llama/Meta-Llama-3.1-8B-Instruct',
          messages: [
            { role: 'system', content: 'You are a traffic analysis expert providing concise, actionable advice.' },
            { role: 'user', content: prompt },
          ],
          max_tokens: 300,
          temperature: 0.7,
        },
        {
          headers: {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json',
          },
        }
      );

      const aiResponse = response.data.choices[0].message.content;
      
      // Parse AI response (simplified)
      const lines = aiResponse.split('\n').filter((l: string) => l.trim());
      
      return {
        analysis: lines[0] || 'Traffic analysis completed',
        recommendations: lines.slice(1, 4),
        severity: this.extractSeverity(aiResponse, situation.averageSpeed),
      };
    } catch (error) {
      logger.error('AI analysis error:', error);
      
      // Fallback to rule-based analysis
      return this.fallbackAnalysis(situation);
    }
  }

  async getRouteRecommendations(
    origin: { lat: number; lon: number },
    destination: { lat: number; lon: number },
    currentTrafficData: any[]
  ): Promise<string[]> {
    try {
      const prompt = `Given traffic data between origin (${origin.lat}, ${origin.lon}) and destination (${destination.lat}, ${destination.lon}), provide 3 alternative route recommendations considering:
- Current congestion levels
- Typical travel times
- Road conditions

Keep recommendations brief and actionable.`;

      const response = await axios.post(
        `${this.baseURL}/chat/completions`,
        {
          model: 'meta-llama/Meta-Llama-3.1-8B-Instruct',
          messages: [
            { role: 'system', content: 'You are a navigation assistant providing optimal route suggestions.' },
            { role: 'user', content: prompt },
          ],
          max_tokens: 200,
          temperature: 0.6,
        },
        {
          headers: {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json',
          },
        }
      );

      const aiResponse = response.data.choices[0].message.content;
      return aiResponse.split('\n').filter((l: string) => l.trim()).slice(0, 3);
    } catch (error) {
      logger.error('Route recommendation error:', error);
      return [
        'Use main highway route (fastest)',
        'Consider alternate route via residential areas',
        'Wait 15-30 minutes for traffic to clear',
      ];
    }
  }

  private extractSeverity(text: string, speed: number): string {
    const lowerText = text.toLowerCase();
    if (lowerText.includes('critical') || lowerText.includes('severe')) return 'critical';
    if (lowerText.includes('high') || speed < 15) return 'high';
    if (lowerText.includes('moderate') || speed < 30) return 'moderate';
    return 'low';
  }

  private fallbackAnalysis(situation: TrafficSituation): AIAnalysisResult {
    const { averageSpeed, vehicleCount } = situation;
    
    let analysis = '';
    let severity: string = 'low';
    const recommendations: string[] = [];

    if (averageSpeed < 10) {
      severity = 'critical';
      analysis = 'Severe congestion detected with very slow traffic movement';
      recommendations.push(
        'Avoid this route if possible',
        'Consider public transportation alternatives',
        'Allow extra 30-45 minutes for travel'
      );
    } else if (averageSpeed < 20) {
      severity = 'high';
      analysis = 'High congestion with significantly reduced speeds';
      recommendations.push(
        'Expect delays of 20-30 minutes',
        'Use alternate routes if available',
        'Consider departure time adjustment'
      );
    } else if (averageSpeed < 35) {
      severity = 'moderate';
      analysis = 'Moderate traffic with some slowdowns';
      recommendations.push(
        'Normal precautions advised',
        'Monitor traffic updates',
        'Expect minor delays'
      );
    } else {
      severity = 'low';
      analysis = 'Traffic flowing smoothly';
      recommendations.push(
        'Good time to travel',
        'Maintain safe following distance',
        'No significant delays expected'
      );
    }

    return { analysis, recommendations, severity };
  }
}

export default new AIService();

File: server/app/services/ai_service.py
"""
Featherless.ai Integration Service
AI-powered traffic insights and recommendations
"""
import httpx
import logging
from typing import Dict, Any, List, Optional
from datetime import datetime

from app.core.config import settings
from app.core.cache import get_cache, set_cache

logger = logging.getLogger(__name__)


class FeatherlessAIService:
    """Featherless.ai API integration for traffic reasoning"""
    
    def __init__(self):
        self.api_key = settings.FEATHERLESS_API_KEY
        self.api_url = settings.FEATHERLESS_API_URL
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
    
    async def analyze_traffic_situation(
        self,
        location_name: str,
        current_congestion: int,
        predicted_congestion: List[Dict[str, Any]],
        historical_patterns: Optional[Dict[str, Any]] = None,
        weather_data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Analyze traffic situation and provide AI insights
        
        Args:
            location_name: Name of the location
            current_congestion: Current congestion level (1-5)
            predicted_congestion: List of predictions
            historical_patterns: Historical traffic patterns
            weather_data: Current weather conditions
        
        Returns:
            AI-generated analysis and recommendations
        """
        # Check cache
        cache_key = f"ai_insight:{location_name}:{datetime.utcnow().strftime('%Y%m%d%H')}"
        cached = await get_cache(cache_key)
        if cached:
            return cached
        
        # Build context for AI
        context = self._build_traffic_context(
            location_name,
            current_congestion,
            predicted_congestion,
            historical_patterns,
            weather_data
        )
        
        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                payload = {
                    "model": "gpt-4",  # or your preferred model
                    "messages": [
                        {
                            "role": "system",
                            "content": "You are an expert traffic analyst. Analyze traffic patterns and provide actionable insights for urban commuters and city planners."
                        },
                        {
                            "role": "user",
                            "content": context
                        }
                    ],
                    "temperature": 0.7,
                    "max_tokens": 500
                }
                
                response = await client.post(
                    f"{self.api_url}/chat/completions",
                    headers=self.headers,
                    json=payload
                )
                
                if response.status_code == 200:
                    result = response.json()
                    analysis = result['choices'][0]['message']['content']
                    
                    insight_data = {
                        'location': location_name,
                        'analysis': analysis,
                        'timestamp': datetime.utcnow().isoformat(),
                        'current_congestion': current_congestion,
                        'predictions': predicted_congestion
                    }
                    
                    # Cache for 1 hour
                    await set_cache(cache_key, insight_data, ttl=3600)
                    
                    return insight_data
                else:
                    logger.error(f"Featherless API error: {response.status_code}")
                    return self._fallback_analysis(location_name, current_congestion, predicted_congestion)
        
        except Exception as e:
            logger.error(f"Featherless AI error: {e}")
            return self._fallback_analysis(location_name, current_congestion, predicted_congestion)
    
    async def get_route_recommendations(
        self,
        origin: str,
        destination: str,
        current_traffic: Dict[str, Any],
        predictions: Dict[str, Any]
    ) -> List[str]:
        """
        Generate AI-powered route recommendations
        
        Args:
            origin: Starting location
            destination: Destination
            current_traffic: Current traffic conditions
            predictions: Traffic predictions
        
        Returns:
            List of route recommendations
        """
        context = f"""
        Route Planning Context:
        - Origin: {origin}
        - Destination: {destination}
        - Current Traffic: {current_traffic}
        - Predicted Traffic: {predictions}
        
        Provide 3 specific route recommendations considering:
        1. Fastest route with current conditions
        2. Best route considering predicted congestion
        3. Alternative route avoiding heavy traffic
        
        Format each recommendation as a bullet point with clear reasoning.
        """
        
        try:
            async with httpx.AsyncClient(timeout=30.0) as client:
                payload = {
                    "model": "gpt-4",
                    "messages": [
                        {
                            "role": "system",
                            "content": "You are a navigation expert. Provide clear, actionable route recommendations."
                        },
                        {
                            "role": "user",
                            "content": context
                        }
                    ],
                    "temperature": 0.7,
                    "max_tokens": 400
                }
                
                response = await client.post(
                    f"{self.api_url}/chat/completions",
                    headers=self.headers,
                    json=payload
                )
                
                if response.status_code == 200:
                    result = response.json()
                    recommendations = result['choices'][0]['message']['content']
                    # Parse into list
                    return [line.strip() for line in recommendations.split('\n') if line.strip()]
                else:
                    return self._fallback_recommendations()
        
        except Exception as e:
            logger.error(f"Route recommendation error: {e}")
            return self._fallback_recommendations()
    
    def _build_traffic_context(
        self,
        location_name: str,
        current_congestion: int,
        predicted_congestion: List[Dict[str, Any]],
        historical_patterns: Optional[Dict[str, Any]],
        weather_data: Optional[Dict[str, Any]]
    ) -> str:
        """Build context string for AI analysis"""
        
        congestion_labels = {1: "Clear", 2: "Light", 3: "Moderate", 4: "Heavy", 5: "Severe"}
        
        context = f"""
        Traffic Analysis Request for {location_name}:
        
        Current Situation:
        - Congestion Level: {current_congestion}/5 ({congestion_labels.get(current_congestion, 'Unknown')})
        - Time: {datetime.utcnow().strftime('%Y-%m-%d %H:%M UTC')}
        
        Predicted Congestion (next 24 hours):
        """
        
        for pred in predicted_congestion[:5]:  # Show top 5
            hours = pred['forecast_hours']
            level = pred['congestion_level']
            context += f"\n- {hours}h ahead: Level {level}/5 ({congestion_labels.get(level, 'Unknown')}), Confidence: {pred['confidence']*100:.0f}%"
        
        if weather_data:
            context += f"\n\nWeather Conditions:\n- Temperature: {weather_data.get('temperature', 'N/A')}°C\n- Conditions: {weather_data.get('condition', 'Unknown')}"
        
        if historical_patterns:
            context += f"\n\nHistorical Pattern: {historical_patterns.get('pattern_description', 'No data')}"
        
        context += """
        
        Please provide:
        1. Brief analysis of the traffic situation
        2. Key factors contributing to current/predicted congestion
        3. Recommendations for commuters (best travel times, alternatives)
        4. Insights for traffic management
        
        Keep the response clear, actionable, and under 300 words.
        """
        
        return context
    
    def _fallback_analysis(
        self,
        location_name: str,
        current_congestion: int,
        predicted_congestion: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Fallback analysis when AI service unavailable"""
        
        congestion_labels = {1: "Clear", 2: "Light", 3: "Moderate", 4: "Heavy", 5: "Severe"}
        
        # Simple rule-based analysis
        if current_congestion >= 4:
            analysis = f"Heavy traffic detected at {location_name}. Consider alternative routes or delaying travel if possible."
        elif current_congestion >= 3:
            analysis = f"Moderate congestion at {location_name}. Expect some delays."
        else:
            analysis = f"Traffic is flowing well at {location_name}. Good time to travel."
        
        # Check predictions
        if predicted_congestion:
            max_future = max([p['congestion_level'] for p in predicted_congestion])
            if max_future >= 4:
                analysis += " Heavy congestion is predicted in the coming hours."
        
        return {
            'location': location_name,
            'analysis': analysis,
            'timestamp': datetime.utcnow().isoformat(),
            'current_congestion': current_congestion,
            'predictions': predicted_congestion,
            'source': 'fallback'
        }
    
    def _fallback_recommendations(self) -> List[str]:
        """Fallback recommendations"""
        return [
            "Check real-time traffic updates before departing",
            "Consider using public transportation during peak hours",
            "Plan your route with alternative options in mind"
        ]
